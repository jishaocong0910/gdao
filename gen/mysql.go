package gen

import (
	"database/sql"
	"fmt"
	"strings"
	"text/template"

	_ "embed"

	_ "github.com/go-sql-driver/mysql"
)

//go:embed mysql_base_dao.tpl
var mysqlBaseDaoTpl string

type mySqlGenerator struct {
	baseDaoTemplate *template.Template
	c               Cfg
	db              *sql.DB
	database        string
}

func (g mySqlGenerator) getBaseDaoTemplate() *template.Template {
	return g.baseDaoTemplate
}

func (g mySqlGenerator) getTableInfo(table string) (bool, []*field, string) {
	var (
		exists       bool
		fields       []*field
		tableComment string
	)

	rows := mustReturn(g.db.Query("SELECT COLUMN_NAME, DATA_TYPE, COLUMN_TYPE, EXTRA = 'auto_increment', COLUMN_COMMENT FROM information_schema.columns WHERE TABLE_SCHEMA = ? AND TABLE_NAME = ? ORDER BY ORDINAL_POSITION", g.database, table))
	defer rows.Close()
	for rows.Next() {
		exists = true
		var (
			column          string
			dataType        string
			columnType      string
			isAutoIncrement bool
			comment         string
		)
		mustNoError(rows.Scan(&column, &dataType, &columnType, &isAutoIncrement, &comment))

		f := &field{
			Column:          column,
			FieldName:       fieldNameMapper.Convert(column),
			FieldType:       "any",
			Comment:         comment,
			IsAutoIncrement: isAutoIncrement,
		}

		dataType = strings.ToLower(dataType)
		columnType = strings.ToLower(columnType)
		switch dataType {
		case "bit":
			f.FieldType = "[]uint8"
		case "tinyint":
			if strings.Contains(columnType, "unsigned") {
				f.FieldType = "*uint8"
			} else if strings.Contains(columnType, "tinyint(1)") {
				f.FieldType = "*bool"
			} else {
				f.FieldType = "*int8"
			}
		case "smallint":
			if strings.Contains(columnType, "unsigned") {
				f.FieldType = "*uint16"
			} else {
				f.FieldType = "*int16"
			}
		case "mediumint":
			if strings.Contains(columnType, "unsigned") {
				f.FieldType = "*uint32"
			} else {
				f.FieldType = "*int32"
			}
		case "int":
			if strings.Contains(columnType, "unsigned") {
				f.FieldType = "*uint32"
			} else {
				f.FieldType = "*int32"
			}
		case "bigint":
			if strings.Contains(columnType, "unsigned") {
				f.FieldType = "*uint64"
			} else {
				f.FieldType = "*int64"
			}
		case "double", "decimal":
			f.FieldType = "*float64"
		case "float":
			f.FieldType = "*float32"
		case "varchar", "char", "text", "tinytext", "mediumtext", "longtext", "enum", "json", "set", "time":
			f.FieldType = "*string"
		case "datetime", "timestamp", "date":
			f.FieldType = "*time.Time"
		case "year":
			f.FieldType = "*int64"
		case "binary", "varbinary", "geometry", "blob", "tinyblob", "mediumblob", "longblob":
			f.FieldType = "[]byte"
		}
		if f.FieldType != "any" {
			f.Valid = true
		}
		fields = append(fields, f)
	}

	rows = mustReturn(g.db.Query("SELECT TABLE_COMMENT FROM information_schema.tables WHERE TABLE_SCHEMA = ? AND TABLE_NAME = ?", g.database, table))
	defer rows.Close()
	if rows.Next() {
		rows.Scan(&tableComment)
	}

	return exists, fields, tableComment
}

func newMySqlGenerator(c Cfg) mySqlGenerator {
	db, err := sql.Open("mysql", c.Dsn)
	if err != nil { // coverage-ignore
		panic(fmt.Sprintf("connect db fail, dsn: %s, error: %v", c.Dsn, err))
	}
	database := ""
	rows := mustReturn(db.Query("SELECT DATABASE()"))
	if rows.Next() {
		rows.Scan(&database)
	}
	rows.Close()
	t := mustReturn(template.New("").Parse(mysqlBaseDaoTpl))
	return mySqlGenerator{baseDaoTemplate: t, c: c, db: db, database: database}
}
