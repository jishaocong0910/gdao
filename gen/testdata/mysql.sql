CREATE TABLE mysql
(
    `auto_increment`     bigint PRIMARY KEY AUTO_INCREMENT COMMENT 'bigint auto_increment',
    `bit`                bit COMMENT 'bit',
    `tinyint`            tinyint COMMENT 'tinyint',
    `tinyint_unsigned`   tinyint UNSIGNED COMMENT 'tinyint unsigned',
    `tinyint_1`          tinyint(1) COMMENT 'tinyint(1)',
    `smallint`           smallint COMMENT 'smallint',
    `smallint_unsigned`  smallint UNSIGNED COMMENT 'smallint unsigned',
    `mediumint`          mediumint COMMENT 'mediumint',
    `mediumint_unsigned` mediumint UNSIGNED COMMENT 'mediumint unsigned',
    `int`                int COMMENT 'int',
    `int_unsigned`       int UNSIGNED COMMENT 'int unsigned',
    `bigint`             bigint COMMENT 'bigint',
    `bigint_unsigned`    bigint UNSIGNED COMMENT 'bigint unsigned',
    `double`             double COMMENT 'double',
    `decimal`            decimal COMMENT 'decimal',
    `varchar`            varchar(255) COMMENT 'varchar(255)',
    `char`               char COMMENT 'char',
    `text`               text COMMENT 'text',
    `tinytext`           tinytext COMMENT 'tinytext',
    `mediumtext`         mediumtext COMMENT 'mediumtext',
    `longtext`           longtext COMMENT 'longtext',
    `enum`               enum ('a', 'b', 'c') COMMENT 'enum(''a'',''b'',''c'')',
    `json`               json COMMENT 'json',
    `set`                set ('a', 'b', 'c') COMMENT 'set(''a'',''b'',''c'')',
    `date`               date COMMENT 'date',
    `datetime`           datetime(3) COMMENT 'datetime(3) ',
    `timestamp`          timestamp COMMENT 'timestamp',
    `year`               year COMMENT 'year',
    `binary`             binary COMMENT 'binary',
    `varbinary`          varbinary(255) COMMENT 'varbinary(255)',
    `geometry`           geometry COMMENT 'geometry',
    `blob`               blob COMMENT 'blob',
    `tinyblob`           tinyblob COMMENT 'tinyblob',
    `mediumblob`         mediumblob COMMENT 'mediumblob',
    `longblob`           longblob COMMENT 'longblob',
    `numeric`            numeric COMMENT 'numeric',
    `integer`            integer COMMENT 'integer',
    `float`              float COMMENT 'float',
    `real`               real COMMENT 'real',
    `boolean`            boolean COMMENT 'boolean',
    `time`               time,
    `other`              int,
    `other2`            int,
    `other3`            int,
    `other4`            int,
    `other5`            int
) COMMENT 'mysql';