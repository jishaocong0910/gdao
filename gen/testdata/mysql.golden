// Code generated by https://github.com/alertego/gdao/gen. YOU CAN EDIT FOR MORE.

package testdata

import "time"

// Mysql mysql
type Mysql struct {
	// AutoIncrement bigint auto_increment
	AutoIncrement *int64 `gdao:"column=auto_increment;auto"`
	// Bit bit
	Bit []uint8 `gdao:"column=bit"`
	// Tinyint tinyint
	Tinyint *int8 `gdao:"column=tinyint"`
	// TinyintUnsigned tinyint unsigned
	TinyintUnsigned *uint8 `gdao:"column=tinyint_unsigned"`
	// Tinyint1 tinyint(1)
	Tinyint1 *bool `gdao:"column=tinyint_1"`
	// Smallint smallint
	Smallint *int16 `gdao:"column=smallint"`
	// SmallintUnsigned smallint unsigned
	SmallintUnsigned *uint16 `gdao:"column=smallint_unsigned"`
	// Mediumint mediumint
	Mediumint *int32 `gdao:"column=mediumint"`
	// MediumintUnsigned mediumint unsigned
	MediumintUnsigned *uint32 `gdao:"column=mediumint_unsigned"`
	// Int int
	Int *int32 `gdao:"column=int"`
	// IntUnsigned int unsigned
	IntUnsigned *uint32 `gdao:"column=int_unsigned"`
	// Bigint bigint
	Bigint *int64 `gdao:"column=bigint"`
	// BigintUnsigned bigint unsigned
	BigintUnsigned *uint64 `gdao:"column=bigint_unsigned"`
	// Double double
	Double *float64 `gdao:"column=double"`
	// Decimal decimal
	Decimal *float64 `gdao:"column=decimal"`
	// Varchar varchar(255)
	Varchar *string `gdao:"column=varchar"`
	// Char char
	Char *string `gdao:"column=char"`
	// Text text
	Text *string `gdao:"column=text"`
	// Tinytext tinytext
	Tinytext *string `gdao:"column=tinytext"`
	// Mediumtext mediumtext
	Mediumtext *string `gdao:"column=mediumtext"`
	// Longtext longtext
	Longtext *string `gdao:"column=longtext"`
	// Enum enum('a','b','c')
	Enum *string `gdao:"column=enum"`
	// Json json
	Json *string `gdao:"column=json"`
	// Set set('a','b','c')
	Set *string `gdao:"column=set"`
	// Date date
	Date *time.Time `gdao:"column=date"`
	// Datetime datetime(3)
	Datetime *time.Time `gdao:"column=datetime"`
	// Timestamp timestamp
	Timestamp *time.Time `gdao:"column=timestamp"`
	// Year year
	Year *int32 `gdao:"column=year"`
	// Binary binary
	Binary []byte `gdao:"column=binary"`
	// Varbinary varbinary(255)
	Varbinary []byte `gdao:"column=varbinary"`
	// Geometry geometry
	Geometry []byte `gdao:"column=geometry"`
	// Blob blob
	Blob []byte `gdao:"column=blob"`
	// Tinyblob tinyblob
	Tinyblob []byte `gdao:"column=tinyblob"`
	// Mediumblob mediumblob
	Mediumblob []byte `gdao:"column=mediumblob"`
	// Longblob longblob
	Longblob []byte `gdao:"column=longblob"`
	// Numeric numeric
	Numeric *float64 `gdao:"column=numeric"`
	// Integer integer
	Integer *int32 `gdao:"column=integer"`
	// Float float
	Float *float32 `gdao:"column=float"`
	// Real real
	Real *float64 `gdao:"column=real"`
	// Boolean boolean
	Boolean *bool `gdao:"column=boolean"`
	// Time
	Time *time.Time `gdao:"column=time"`
	// Other
	Other []string `gdao:"column=other"`
	// Other2
	// gdao cannot solve this type!
	Other2 []rune `gdao:"column=other2"`
	// Other3
	Other3 *string `gdao:"column=other3"`
	// Other4
	Other4 *string `gdao:"column=other4"`
	// Other5
	// gdao cannot solve this type!
	Other5 *rune `gdao:"column=other5"`
}
